api/utils/sshutils/ssh.go

```diff
@@ -250,3 +260,263 @@ func KeysEqual(ak, bk ssh.PublicKey) bool {
+	teleportPublicKeyDecoded := make([]byte, base64.StdEncoding.EncodedLen(len(teleportPublicKeyBytes)))
+	teleportPublicKeyByteCount, err := base64.StdEncoding.Decode(teleportPublicKeyDecoded, teleportPublicKeyBytes)
+	publicKeyData := make([]byte, teleportPublicKeyByteCount)
+	copy(publicKeyData, teleportPublicKeyDecoded)
+	if err != nil {
+		return nil, trace.Wrap(fmt.Errorf("could not base64-decode public key: %T, got %v bytes successfully", err, teleportPublicKeyByteCount))
+	}
+	// append the decoded public key bytes to the MAC buffer
+	macPublicKeyData, err := getRFC4251String(publicKeyData)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	err = binary.Write(macInput, binary.BigEndian, macPublicKeyData)
+	if err != nil {
+		return nil, trace.Wrap(fmt.Errorf("cannot write macPublicKeyData to buffer: %T", err))
+	}
+
+	// append our PPK-formatted private key bytes to the MAC buffer
+	macPrivateKeyData, err := getRFC4251String(ppkPrivateKey.Bytes())
+	if err != nil {
+		return nil, err
+	}
+	err = binary.Write(macInput, binary.BigEndian, macPrivateKeyData)
+	if err != nil {
+		return nil, trace.Wrap(fmt.Errorf("cannot write macPrivateKeyData to buffer: %T", err))
+	}
+
+	// as per the PPK spec, the key for the MAC is blank when the PPK file is unencrypted.
+	// therefore, the key is a zero-length byte slice.
+	hmacHash := hmac.New(sha256.New, []byte{})
+	// generate the MAC using HMAC-SHA-256
+	hmacHash.Write(macInput.Bytes())
+	macString := hex.EncodeToString(hmacHash.Sum(nil))
+
+	// build the string-formatted output PPK file
+	var ppk strings.Builder
+	ppk.WriteString(fmt.Sprintf("PuTTY-User-Key-File-3: %v\n", keyType))
+	ppk.WriteString(fmt.Sprintf("Encryption: %v\n", encryptionType))
+	ppk.WriteString(fmt.Sprintf("Comment: %v\n", fileComment))
+	// chunk the Teleport-formatted public key into 64-character length lines
+	chunkedPublicKey := chunkString(teleportPublicKey, 64)
+	ppk.WriteString(fmt.Sprintf("Public-Lines: %v\n", len(chunkedPublicKey)))
+	for _, r := range chunkedPublicKey {
+		ppk.WriteString(fmt.Sprintf("%s\n", r))
+	}
+	// chunk the PPK-formatted private key into 64-character length lines
+	chunkedPrivateKey := chunkString(string(ppkPrivateKeyBase64), 64)
+	ppk.WriteString(fmt.Sprintf("Private-Lines: %v\n", len(chunkedPrivateKey)))
+	for _, r := range chunkedPrivateKey {
+		ppk.WriteString(fmt.Sprintf("%s\n", r))
+	}
+	ppk.WriteString(fmt.Sprintf("Private-MAC: %v\n", macString))
+
+	// convert string to bytes and return
+	return []byte(ppk.String()), nil
+}
+
+// chunkString converts a string into a []string with chunks of size chunkSize
+// used to split base64-encoded strings across multiple lines with an even width
+func chunkString(s string, chunkSize int) []string {
+	if len(s) == 0 {
+		return nil
+	}
+	if chunkSize >= len(s) {
+		return []string{s}
+	}
+	chunks := make([]string, 0, (len(s)-1)/chunkSize+1)
+	currentLen := 0
+	currentStart := 0
+	for i := range s {
```

Comment:
`range` over a string iterates over Unicode code points, not bytes. If this function is supposed to produce chunks of a particular number of bytes, then we should take a different approach.

Reply:
I get the distinction. Given that these values will always be base64'd (so the code points will only ever be in the ASCII set) I think we can stick with this approach?

I'll add a warning that it's not multi-byte safe.